eliot/_parse.py

```diff
@@ -0,0 +1,97 @@
+from ._action import (
+    TaskLevel, WrittenAction, ACTION_STATUS_FIELD, STARTED_STATUS,
+    ACTION_TYPE_FIELD,
+)
+
+
+class Task(PClass):
+    """
+    A tree of actions with the same task UUID.
+    """
+    _nodes = pmap_field(TaskLevel, (WrittenAction, WrittenMessage))
+
+    _root_level = TaskLevel(level=[])
+
+    def root(self):
+        """
+        @return: The root L{WrittenAction}.
+        """
+        return self._nodes[self._root_level]
+
+    def _insert_action(self, node):
+        """
+        Add a L{WrittenAction} to the tree.
+
+        Parent actions will be created as necessary.
+
+        @param child: A L{WrittenAction} to add to the tree.
+
+        @return: Updated L{Task}.
+        """
+        task = self.transform(["_nodes", node.task_level], node)
+        return task._ensure_node_parents(node)
+
+    def _ensure_node_parents(self, child):
+        """
+        Ensure the node (WrittenAction/WrittenMessage) is referenced by parent
+        nodes.
+
+        Parent actions will be created as necessary.
+
+        @param child: A L{WrittenMessage} or L{WrittenAction} which is
+            being added to the tree.
+
+        @return: Updated L{Task}.
+        """
+        task_level = child.task_level
+        if task_level.parent() is None:
+            return self
+
+        parent = self._nodes.get(task_level.parent())
+        if parent is None:
+            parent = WrittenAction(task_level=task_level.parent(),
+                                   task_uuid=child.task_uuid)
+        parent = parent._add_child(child)
+        return self._insert_action(parent)
+
+    def add(self, message_dict):
+        """
+        Update the L{Task} with a dictionary containing a serialized Eliot
+        message.
+
+        @param message_dict: Dictionary whose task UUID matches this one.
+
+        @return: Updated L{Task}.
+        """
+        is_action = message_dict.get(ACTION_TYPE_FIELD) is not None
+        written_message = WrittenMessage.from_dict(message_dict)
+        if is_action:
+            action_level = written_message.task_level.parent()
+            action = self._nodes.get(action_level)
+            if action is None:
+                action = WrittenAction(task_level=action_level,
+                                       task_uuid=message_dict[TASK_UUID_FIELD])
+            if message_dict[ACTION_STATUS_FIELD] == STARTED_STATUS:
+                # Either newly created MissingAction, or one created by
+                # previously added descendant of the action.
+                action = action._start(written_message)
+            else:
+                action = action._end(written_message)
+            return self._insert_action(action)
+        else:
+            # Special case where there is no action:
+            if written_message.task_level.level == [1]:
+                return self.transform(
+                    ["_nodes", self._root_level], written_message)
```

Comment:
I'll double check this in a bit.

My reading is that you can have messages without any actions, in which case you get a task that only has a single node being the WrittenMessage itself.

Expect another comment from me once I confirm that's true.